### PR TITLE
New tainting_rule type and parse its yaml file

### PR DESCRIPTION
**Pull request by aryx**
This is the first step towards taint-tracking integration into semgrep

This will help issue #585

Test plan:
/home/pad/github/semgrep/semgrep-core/_build/default/bin/main_semgrep_core.exe -lang js -dump_tainting_rules data/basic_tainting.yml
[("classic", [Tag3 (Tag10 (Tag7 (["source"; (("source", 0, 1, 0, "/tmp/tmp-29556-3308e5.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 7, 1, 7, "/tmp/tmp-29556-3308e5.py"))]))])); Tag3 (Tag10 (Tag7 (["source1"; (("source1", 0, 1, 0, "/tmp/tmp-29556-610dbd.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 8, 1, 8, "/tmp/tmp-29556-610dbd.py"))]))]))], [Tag3 (Tag10 (Tag7 (["sanitize"; (("sanitize", 0, 1, 0, "/tmp/tmp-29556-ad5a50.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 9, 1, 9, "/tmp/tmp-29556-ad5a50.py"))]))]))], [Tag3 (Tag10 (Tag7 (["sink"; (("sink", 0, 1, 0, "/tmp/tmp-29556-104d40.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 5, 1, 5, "/tmp/tmp-29556-104d40.py"))]))])); Tag3 (Tag10 (Tag7 (["eval"; (("eval", 0, 1, 0, "/tmp/tmp-29556-4b39bc.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 5, 1, 5, "/tmp/tmp-29556-4b39bc.py"))]))]))], "A user input source() went into a dangerous sink()", 1, [0; 3])]

### CODE REVIEW

File: semgrep-core/core/tainting_rule.ml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+(* right now only Expr is supported *)
+type pattern = Ast.any
+
+(* less: could extend Rule.t *)
+type rule = {
+  id: string;
+
+  (* the list below are used to express disjunction *)
+  source: pattern list;
```
Comment (DrewDennison) [inline semgrep-core/core/tainting_rule.ml:38]: nice!

**Review by DrewDennison — APPROVED** [review]
LGTM

File: semgrep-core/data/basic_tainting.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+rules:
+  - id: classic
+    pattern-sources:
+      - source(...)
+      - source1(...)
+    pattern-sinks:
+      - sink(...)
+      - eval(...)
```
Comment (ulziibay) [inline semgrep-core/data/basic_tainting.yml]: Just to be sure, these are logical OR clauses. In addition to this, is it possible to define two sinks combined by AND clause?

For example, when we were doing path-injection, we had to make patterns like
```
$P = os.path.join(..., request.$W.get(...), ...)
...
open($P, ...)
```
where data had to flow through `os.path.join` as well as `open` to be considered potentially malicious.

**Review by ulziibay — COMMENTED** [review]
the yaml API makes sense to me. Just left one additional request

File: semgrep-core/data/basic_tainting.yml (hunk 2 above)
Comment (aryx) [reply semgrep-core/data/basic_tainting.yml]: hmm, right now the tainting by default consider that any tainted argument taint also the return value (unless it's a sanitize function), so here by default $P will be marked as tainted

**Comment by brendongo** [comment]
@aryx how do the sinks/sanitizers/sources interact with other rule.yaml keys like pattern-either/pattern/pattern-inside etc.

**Comment by aryx** [comment]
It's in semgrep-core, which does not know anything about the logic-based pattern combination.
Right now semgrep transforms the complex yaml file (with logic operators) into a flat list of simple patterns for semgrep-core in a simpler yaml file (see semgrep-core/core/rule.ml). We could imagine the same for the tainted rules. @nbrahms had some ideas about that.

**Comment by aryx** [comment]
We probably need a meeting to figure things out. I've created this tainted rule yaml file format as a spin-off, just to get quickly to the core of the problem. In the future I would prefer to have it nicely integrated with the rest of semgrep, to have something like $X: tainted metavariable, or something, but it requires some language design thinking.

**Comment by aryx** [comment]
Right now I just imagine a separate tab in semgrep.live: we have the single pattern tab, then yaml logic-based operators config file, and I would imagine another yaml tainting config file tab. That way we can quickly give this to grayson/matt/ulzii/... to play with and get feedback on where the analysis is good and better than the logic-based patterns we wrote before and where we can improve.

**Pull request merged**